byoqm/source_repository/source_repository.py

```diff
@@ -14,55 +24,103 @@ class SourceRepository:
+        get_ast checks if an AST for the path already has been computed, and returns that
         AST if it is the case. Otherwise, it will parse the file to a tree_sitter AST
         and return that tree.
-
-        Throws a ValueError when the file to parse is not a child path of the given
-        src_root.
         """
-        if for_file not in self.src_paths:
-            logging.error("The file to parse must be a child path of of the src_root")
-            raise ValueError
-
         ast = None
         try:
-            ast = self.asts[for_file]
+            ast = self.asts[for_file.file_path]
+            return ast
         except KeyError:
-            self.asts[for_file] = self._parse_ast(for_file)
-            ast = self.asts[for_file]
-        finally:
+            self.asts[for_file.file_path] = self._parse_ast(for_file)
+            ast = self.asts[for_file.file_path]
             return ast
 
-    def _parse_ast(self, file_at: Path) -> tree_sitter.Tree:
+    def _parse_ast(self, of_file: FileInfo) -> tree_sitter.Tree:
         """
         parses and returns the tree_sitter AST for a given file
         """
-        try:
-            with file_at.open("rb") as file:
-                ast = self._parser.parse(file.read())
-                return ast
-        except Exception as e:
-            logging.error(
-                f"Failed to parse ast for file at path: {file_at} with encoding {self.file_encodings[file_at]}. Error: {e}"
-            )
-            raise e
+        with of_file.file_path.open("rb") as file:
+            ast = self.tree_sitter_parsers[of_file.language].parse(file.read())
+            if ast is None:
+                raise TypeError(f"Abstract syntax tree for ${of_file} is None")
+            return ast
+
+    def _discover_files(self) -> Dict[Path, FileInfo]:
+        if self.src_root.is_file():
+            return {self.src_root: self._inspect_file(self.src_root)}
+
+        file_infos: Dict[Path, FileInfo] = self._discover_in_dir(self.src_root)
+        return file_infos
+
+    def _discover_in_dir(self, root_dir: Path) -> Dict[Path, FileInfo]:
+        file_infos: Dict[Path, FileInfo] = {}
+        for f in root_dir.iterdir():
+            if f.is_dir():
+                file_infos.update(self._discover_in_dir(f))
+            else:
+                file_info = self._inspect_file(f)
+                if not self._should_exclude(file_info):
+                    file_infos[f] = file_info
+
+        return file_infos
+
+    def _inspect_file(self, file_path: Path) -> FileInfo:
+        if not file_path.is_file():
+            raise ValueError(f"_inspect_file expects that ${file_path} is a file")
+
+        programming_language: str = "unknown"
+        if file_path.suffix == ".py":
+            programming_language = "python"
+        elif file_path.suffix == ".cs":
+            programming_language = "c_sharp"
+        elif file_path.suffix == ".java":
+            programming_language = "java"
```

Comment:
we need some error handling on this. Either a default else case or something similar that will stop the program if an error is encountered

Reply:
Unknown isn't an immediate error. It just means that we have a file that we don't want to analyze because it is not one of c#, java, or python, so we want to exclude it from the list of files in the source repository,